Dict unpacking within a context manager triggers `missing-kwoa`

### Bug description

Reproduction:
```python
import contextlib

@contextlib.contextmanager
def f(*, a):
    yield
    
def g(**kw):
    f(**kw)  # doesn't trigger missing-kwoa
    with f(**kw):  # triggers missing-kwoa
        pass
        
g(a=1)
```

Seems related to #1111

### Configuration

_No response_

### Command used

```shell
pylint repro.py
```


### Pylint output

```shell
$ pylint repro.py
************* Module repro
repro.py:6:0: C0303: Trailing whitespace (trailing-whitespace)
repro.py:11:0: C0303: Trailing whitespace (trailing-whitespace)
repro.py:1:0: C0114: Missing module docstring (missing-module-docstring)
repro.py:4:0: C0116: Missing function or method docstring (missing-function-docstring)
repro.py:4:0: C0103: Function name "f" doesn't conform to snake_case naming style (invalid-name)
repro.py:4:9: W0613: Unused argument 'a' (unused-argument)
repro.py:7:0: C0116: Missing function or method docstring (missing-function-docstring)
repro.py:7:0: C0103: Function name "g" doesn't conform to snake_case naming style (invalid-name)
repro.py:9:9: E1125: Missing mandatory keyword argument 'a' in function call (missing-kwoa)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```


### Expected behavior

`missing-kwoa` shouldn't be triggered

### Pylint version

```shell
pylint 2.15.10
astroid 2.13.3
Python 3.10.9 (tags/v3.10.9:1dd9be6, Dec  6 2022, 20:01:21) [MSC v.1934 64 bit (AMD64)]
```


### OS / Environment

_No response_

### Additional dependencies

_No response_

--- a/doc/whatsnew/fragments/8100.false_positive
+++ b/doc/whatsnew/fragments/8100.false_positive
@@ -1,0 +1,3 @@
+Fix ``no-kwoa`` false positive for context managers.
+
+Closes #8100

--- a/pylint/checkers/typecheck.py
+++ b/pylint/checkers/typecheck.py
@@ -691,13 +691,7 @@ def _no_context_variadic_keywords(node: nodes.Call, scope: nodes.Lambda) -> bool
     ):
         call = statement.value
         variadics = list(call.keywords or []) + call.kwargs
-    # elif isinstance(statement, nodes.With):
-    #     breakpoint()
-    #     calls = statement.items
-    #     if calls:
-    #         # only handle one call for now...
-    #         call = statement.items[0][0]
-    #         variadics = list(call.keywords or []) + call.kwargs
+
     return _no_context_variadic(node, scope.args.kwarg, nodes.Keyword, variadics)
 
 
@@ -1622,7 +1616,12 @@ accessed. Python regular expressions are accepted.",
                 and not has_no_context_keywords_variadic
                 and not overload_function
             ):
-                self.add_message("missing-kwoa", node=node, args=(name, callable_name))
+                self.add_message(
+                    "missing-kwoa",
+                    node=node,
+                    args=(name, callable_name),
+                    confidence=INFERENCE,
+                )
 
     @staticmethod
     def _keyword_argument_is_in_all_decorator_returns(

--- a/pylint/checkers/typecheck.py
+++ b/pylint/checkers/typecheck.py
@@ -687,7 +687,12 @@ def _no_context_variadic_keywords(node: nodes.Call, scope: nodes.Lambda) -> bool
     ):
         call = statement.value
         variadics = list(call.keywords or []) + call.kwargs
-
+    elif isinstance(statement, nodes.With):
+        calls = statement.items
+        if calls:
+            # only handle one call for now...
+            call = statement.items[0][0]
+            variadics = list(call.keywords or []) + call.kwargs
     return _no_context_variadic(node, scope.args.kwarg, nodes.Keyword, variadics)
 
 

--- a/pylint/checkers/typecheck.py
+++ b/pylint/checkers/typecheck.py
@@ -680,19 +680,24 @@ def _no_context_variadic_keywords(node: nodes.Call, scope: nodes.Lambda) -> bool
     statement = node.statement(future=True)
     variadics = []
 
-    if isinstance(scope, nodes.Lambda) and not isinstance(scope, nodes.FunctionDef):
+    if (
+        isinstance(scope, nodes.Lambda)
+        and not isinstance(scope, nodes.FunctionDef)
+        or isinstance(statement, nodes.With)
+    ):
         variadics = list(node.keywords or []) + node.kwargs
     elif isinstance(statement, (nodes.Return, nodes.Expr, nodes.Assign)) and isinstance(
         statement.value, nodes.Call
     ):
         call = statement.value
         variadics = list(call.keywords or []) + call.kwargs
-    elif isinstance(statement, nodes.With):
-        calls = statement.items
-        if calls:
-            # only handle one call for now...
-            call = statement.items[0][0]
-            variadics = list(call.keywords or []) + call.kwargs
+    # elif isinstance(statement, nodes.With):
+    #     breakpoint()
+    #     calls = statement.items
+    #     if calls:
+    #         # only handle one call for now...
+    #         call = statement.items[0][0]
+    #         variadics = list(call.keywords or []) + call.kwargs
     return _no_context_variadic(node, scope.args.kwarg, nodes.Keyword, variadics)
 
 

--- a/tests/functional/d/dataclass/dataclass_kw_only.txt
+++ b/tests/functional/d/dataclass/dataclass_kw_only.txt
@@ -1,4 +1,4 @@
-missing-kwoa:24:0:26:1::Missing mandatory keyword argument 'a' in constructor call:UNDEFINED
-missing-kwoa:24:0:26:1::Missing mandatory keyword argument 'b' in constructor call:UNDEFINED
+missing-kwoa:24:0:26:1::Missing mandatory keyword argument 'a' in constructor call:INFERENCE
+missing-kwoa:24:0:26:1::Missing mandatory keyword argument 'b' in constructor call:INFERENCE
 redundant-keyword-arg:24:0:26:1::Argument 'c' passed by position and keyword in constructor call:UNDEFINED
 too-many-function-args:24:0:26:1::Too many positional arguments for constructor call:UNDEFINED

--- a/tests/functional/m/missing/missing_kwoa.py
+++ b/tests/functional/m/missing/missing_kwoa.py
@@ -86,5 +86,7 @@ def test_context_managers(**kw):
     with run(**kw), run(**kw):
         pass
 
+    with run(**kw), run():  # [missing-kwoa]
+        pass
 
 test_context_managers(a=1)

--- a/tests/functional/m/missing/missing_kwoa.py
+++ b/tests/functional/m/missing/missing_kwoa.py
@@ -1,6 +1,6 @@
 # pylint: disable=missing-docstring,unused-argument,too-few-public-methods
+import contextlib
 import typing
-
 
 def target(pos, *, keyword):
     return pos + keyword
@@ -71,3 +71,20 @@ class Child(Parent):
             second):
         super().__init__(first=first, second=second)
         self._first = first + second
+
+
+@contextlib.contextmanager
+def run(*, a):
+    yield
+
+def test_context_managers(**kw):
+    run(**kw)
+
+    with run(**kw):
+        pass
+
+    with run(**kw), run(**kw):
+        pass
+
+
+test_context_managers(a=1)

--- a/tests/functional/m/missing/missing_kwoa.txt
+++ b/tests/functional/m/missing/missing_kwoa.txt
@@ -1,3 +1,4 @@
 missing-kwoa:21:4:21:17:not_forwarding_kwargs:Missing mandatory keyword argument 'keyword' in function call:UNDEFINED
 missing-kwoa:27:0:27:16::Missing mandatory keyword argument 'keyword' in function call:UNDEFINED
 too-many-function-args:27:0:27:16::Too many positional arguments for function call:UNDEFINED
+missing-kwoa:89:20:89:25:test_context_managers:Missing mandatory keyword argument 'a' in function call:UNDEFINED

--- a/tests/functional/m/missing/missing_kwoa.txt
+++ b/tests/functional/m/missing/missing_kwoa.txt
@@ -1,4 +1,4 @@
-missing-kwoa:21:4:21:17:not_forwarding_kwargs:Missing mandatory keyword argument 'keyword' in function call:UNDEFINED
-missing-kwoa:27:0:27:16::Missing mandatory keyword argument 'keyword' in function call:UNDEFINED
+missing-kwoa:21:4:21:17:not_forwarding_kwargs:Missing mandatory keyword argument 'keyword' in function call:INFERENCE
+missing-kwoa:27:0:27:16::Missing mandatory keyword argument 'keyword' in function call:INFERENCE
 too-many-function-args:27:0:27:16::Too many positional arguments for function call:UNDEFINED
-missing-kwoa:89:20:89:25:test_context_managers:Missing mandatory keyword argument 'a' in function call:UNDEFINED
+missing-kwoa:89:20:89:25:test_context_managers:Missing mandatory keyword argument 'a' in function call:INFERENCE

